README
MooseFS is an Open Source, easy to deploy and maintain, distributed,
fault tolerant file system for POSIX compliant OSes.

MooseFS uses FUSE (http://fuse.sourceforge.net/).

============
Date of first public release: 2008-05-30

Project web site: http://www.moosefs.com/

Installation and using MooseFS:
http://www.moosefs.org/reference-guide.html

Sourceforge project site with source repository:
http://sourceforge.net/projects/moosefs/

Reporting bugs: bugs@moosefs.com
General contact address: contact@moosefs.com

Copyright
=========

Copyright Jakub Kruszona-Zawadzki, Core Technology Sp. z o.o.

READ THIS BEFORE INSTALLING THE SOFTWARE. BY INSTALLING,
ACTIVATING OR USING THE SOFTWARE, YOU ARE AGREEING TO BE BOUND BY
THE TERMS AND CONDITIONS OF MooseFS LICENSE AGREEMENT FOR
VERSION 1.7 AND HIGHER IN A SEPARATE FILE. THIS SOFTWARE IS LICENSED AS
THE PROPRIETARY SOFTWARE, NOT AS OPEN SOURCE ONE. YOU NOT ACQUIRE
ANY OWNERSHIP RIGHT, TITLE OR INTEREST IN OR TO ANY INTELLECTUAL
PROPERTY OR OTHER PROPRITARY RIGHTS.


NEWS
This file lists noteworthy changes in MooseFS.

* MooseFS 2.0.54-1 (2015-03-11)

  - (mount) fixed network timeouts in master-mount connection module.

* MooseFS 2.0.53-1 (2015-03-10)

  - (master) fixed possible reference to freed memory during metadata recovery
  - (master) added minimal priority queue length
  - (cs) fixed removing damaged disks (lock file issue)
  - (master) treat servers in maintenance mode as overloaded (no chunk creations, replications etc.)

* MooseFS 2.0.52-1 (2015-03-04)

  - (master) added option for setting length of chunks priority queues

* MooseFS 2.0.51-1 (2015-02-26)

  - (master+cgi+cli) added quota-licences (pro version only)

* MooseFS 2.0.50-1 (2015-02-10)

  - (master) fixed descriptor leak

* MooseFS 2.0.49-1 (2015-02-02)

  - (cs) made fsync before close optional
  - (all) added missing options to manpages

* MooseFS 2.0.48-1 (2015-01-14)

  - (cs) fixed error handling in writing data (mainserv)
  - (mount) fixed error handling bug in writing data module

* MooseFS 2.0.47-1 (2015-01-09)

  - (cs) fixed handling .metaid file
  - (cs) fixed showing incorrect I/O error messages
  - (cs) fixed drive removals

* MooseFS 2.0.46-1 (2015-01-08)

  - (master) force rebalancing when some servers are almost full (regardless of undergoal priority queues)
  - (master) fixed potential double decreasing number of sockets in session data

* MooseFS 2.0.45-1 (2014-12-24)

  - (mount) fixed printed informations about mapall (thanks to Tom Ivar Helbekkmo)
  - (master) fixed rmdir mapall/maproot rights check (thanks to Tom Ivar Helbekkmo)

* MooseFS 2.0.44-1 (2014-11-27)

  - (master) fixed read rebalance limit

* MooseFS 2.0.43-1 (2014-10-31)

  - (master) introducing new licence model (new licences with size limit for pro version)
  - (all) fixed portability issues with usleep (thanks to Tom Ivar Helbekkmo)
  - (all) setting socket options after listen (thanks to Tom Ivar Helbekkmo)

* MooseFS 2.0.42-1 (2014-10-29)

  - (tools) fixed printing ip numbers in mfsfileinfo
  - (mount) changed format of ip numbers in messages from hex to standard
  - (master) changed format of ip numbers in some messages from hex to standard

* MooseFS 2.0.41-1 (2014-10-21)

  - (mount) fixed buffer overrun (FreeBSD and OS X only)

* MooseFS 2.0.40-1 (2014-10-15)

  - (cs) fixed serious bug (new chunks were created on disks 'marked for removal') - introduced in 1.7.31
  - (cs) fixed bug (segfault on very rare ocassions during disk scanning just after startup - caused by damaged chunk files)
  - (mount) fixed memory leak (supplementary group tabs) - intr. in 2.0.1
  - (rpm) fixed systemd scripts
  - (daemons) added increasing limit of core dump size (debugging purposes)

* MooseFS 2.0.39-1 (2014-10-06)

  - (mount) added ability to mount from /etc/fstab on FreeBSD

* MooseFS 2.0.38-1 (2014-10-02)

  - (mount) fixed memory leak in workaround on FreeBSD early-release bug (FreeBSD only)

* MooseFS 2.0.37-1 (2014-09-23)

  - (cs) added internal open files limit
  - (master) changed chunkserver choosing algorithm

* MooseFS 2.0.36-1 (2014-09-18)

  - (cgi) fixed sorting masters by ip
  - (master) fixed acl posix compatibility (two tests from pjd)
  - (cs) fixed compatibility issues with master 1.6

* MooseFS 2.0.35-1 (2014-09-12)

  - (freebsd) fixed cgiserv and cli port
  - (mount) fixed setting acl

* MooseFS 2.0.34-1 (2014-08-21)

  - (mount) workaround for bug in FreeBSD Fuse version
  - (master) added setting correct metaid in "new" followers (pro version only)

* MooseFS 2.0.33-1 (2014-08-18)

  - (cs,supervisor) more reliable network code in supervisor module (pro version only)
  - (cs,master) added check for metaid in chunkservers (also hard drives) and secondary masters

* MooseFS 2.0.32-1 (2014-08-13)

  - (cs) changed delay handling method (delayed file close,free crc block etc.)
  - (mount) fixed truncate loop (fixed bug intr. in 2.0.26)

* MooseFS 2.0.31-1 (2014-08-12)

  - (cs) fixed bug: write data error not causing exit from write loop - sometimes could lead to segfault.
  - (master) fixed setting data cache drop bit.
  - (cli) fixed printed options for mfssupervisor

* MooseFS 2.0.30-1 (2014-07-09)

  - (daemons) added writing pid to lockfile
  - (all) added freebsd port-maker

* MooseFS 2.0.29-1 (2014-07-03)

  - (cs) moved fsync to background (improved write performance)

* MooseFS 2.0.28-1 (2014-06-30)

  - (master) fixed syslog message about metaloggers in masters (pro version only)

* MooseFS 2.0.27-1 (2014-06-27)

  - (tools) added to mfsfileinfo checksum digest and file signature calculation

* MooseFS 2.0.26-1 (2014-06-26)

  - (cs) fixed bug in replicator
  - (mount) fixed truncate loop
  - (mount) fixed potential descriptor leak
  - (master) fixed bug with duplicated chunkserver id

* MooseFS 2.0.25-1 (2014-06-23)

  - (mount) better handling error counters
  - (master) fixed replication status handling

* MooseFS 2.0.24-1 (2014-06-23)

  - (master) added unfinished replications detection
  - (master) added source ip and port in replication status message
  - (cs) fixed handling job buffers
  - (daemons) added ignoring profiling signals

* MooseFS 2.0.23-1 (2014-06-18)

  - (mount) changed chunkserver choosing algorithm
  - (master) added checks for busy/operation mismatch
  - (master) added checks for busy status during duplicate/duplicate+truncate operations
  - (cs) added ignoring 'marked for removal' disks in internal rebalance

* MooseFS 2.0.22-1 (2014-06-17)

  - (master) add chunks to priority queues after chunk damage, chunk lost and set goal

* MooseFS 2.0.21-1 (2014-06-17)

  - (supervisor+cstool) fixed version shown
  - (supervisor) fixed 'DEAD' master recognition routine
  - (master) fixed segfault introduced in 2.0.20

* MooseFS 2.0.20-1 (2014-06-16)

  - (cs) fixed handling unfinished jobs on socket timeout
  - (master) fixed missing chunks CHUNKDEL LEADER/FOLLOWER race condition (pro version only)

mfscommon/md5.c
/*
   Copyright Jakub Kruszona-Zawadzki, Core Technology Sp. z o.o.

   This file is part of MooseFS.

   READ THIS BEFORE INSTALLING THE SOFTWARE. BY INSTALLING,
   ACTIVATING OR USING THE SOFTWARE, YOU ARE AGREEING TO BE BOUND BY
   THE TERMS AND CONDITIONS OF MooseFS LICENSE AGREEMENT FOR
   VERSION 1.7 AND HIGHER IN A SEPARATE FILE. THIS SOFTWARE IS LICENSED AS
   THE PROPRIETARY SOFTWARE. YOU NOT ACQUIRE
   ANY OWNERSHIP RIGHT, TITLE OR INTEREST IN OR TO ANY INTELLECTUAL
   PROPERTY OR OTHER PROPRIETARY RIGHTS.
 */

#include <string.h>
#include <inttypes.h>
#include "md5.h"

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static const uint8_t padding[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

static void md5_encode (uint8_t *output,const uint32_t *input,uint32_t len) {
	uint32_t i,j;

	for (i=0,j=0 ; i<len ; i++,j+=4) {
		output[j] = (uint8_t)(input[i]&0xff);
		output[j+1] = (uint8_t)((input[i]>>8)&0xff);
		output[j+2] = (uint8_t)((input[i]>>16)&0xff);
		output[j+3] = (uint8_t)((input[i]>>24)&0xff);
	}
}

static void md5_decode (uint32_t *output,const uint8_t *input,uint32_t len) {
	uint32_t i,j;

	for (i=0,j=0 ; i<len ; i++,j+=4) {
		output[i] = ((uint32_t)input[j]) | (((uint32_t)input[j+1])<<8) | (((uint32_t)input[j+2])<<16) | (((uint32_t)input[j+3])<<24);
	}
}

static void md5_transform (uint32_t state[4],const uint8_t block[64]) {
	uint32_t a=state[0], b=state[1], c=state[2], d=state[3], x[16];

	md5_decode(x,block,16);

	FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	memset((char*)x,0,sizeof(x));
}



void md5_init (md5ctx *ctx) {
	ctx->count[0] = ctx->count[1] = 0;
	ctx->state[0] = 0x67452301U;
	ctx->state[1] = 0xefcdab89U;
	ctx->state[2] = 0x98badcfeU;
	ctx->state[3] = 0x10325476U;
}

void md5_update (md5ctx *ctx,const uint8_t *buff,uint32_t leng) {
	uint32_t i,indx,partleng;

	indx = (uint32_t)((ctx->count[0]>>3)&0x3F);

	if ((ctx->count[0]+=((uint32_t)leng<<3)) < ((uint32_t)leng<<3)) {
		ctx->count[1]++;
	}
	ctx->count[1] += ((uint32_t)leng>>29);

	partleng = 64-indx;

	if (leng>=partleng) {
		memcpy((char*)(ctx->buffer+indx),(const char*)buff,partleng);
		md5_transform(ctx->state,ctx->buffer);

		for (i=partleng ; i+63<leng ; i+=64) {
			md5_transform(ctx->state,buff+i);
		}
		indx = 0;
	} else {
		i = 0;
	}

	memcpy((char*)(ctx->buffer+indx),(const char*)(buff+i),leng-i);
}

void md5_final (uint8_t digest[16],md5ctx *ctx) {
	uint8_t bits[8];
	uint32_t indx,padleng;

	md5_encode(bits,ctx->count,2);
	indx = (uint32_t)((ctx->count[0]>>3)&0x3f);
	padleng = (indx<56)?(56-indx):(120-indx);
	md5_update (ctx,padding,padleng);
	md5_update(ctx,bits,8);
	md5_encode(digest,ctx->state,4);
	memset((char*)ctx,0,sizeof(md5ctx));
}
